flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -730,12 +730,7 @@ public final void cancel() throws Exception {
                             (unusedResult, unusedError) -> {
                                 // WARN: the method is called from the task thread but the callback
                                 // can be invoked from a different thread
-                                mailboxProcessor.allActionsCompleted();
-                                try {
-                                    cancelables.close();
-                                } catch (IOException e) {
-                                    throw new CompletionException(e);
-                                }
+                                stopMailboxProcessor();
```

Comment:
After your change `stopMailboxProcessor()` has an implicit contract that it should be thread safe and because of that I'm not sure if extracting this code to separate method as it is is a good idea. There is `WARN` comment that you have missed. 

(Previously that was a private implementation detail of this lambda method that it has to be thread safe. Now you are moving it to the `StreamTask` interface)

Reply:
Not sure if I fully understand the concern. Can you help me understand more? Previously it is possible that the logic in `stopMailboxProcessor` is run in a different thread from the mailbox thread, e.g. the `SourceThread` in `SourceStreamTask`. So supposedly the `stopMailboxProcessor` logic should already be thread safe. And expose it to subclasses should also be safe. Is that the correct?

Or do you mean by exposing the method to subclasses, the subclasses may stop the mailbox any any given time without coordinating with the control flow in the `StreamTask`?  But both `mailboxProcessor` and `getCancelables` are already available to subclasses, so the subclasses can do that anyways.

Reply:
I guess the problem partially (?) comes from the de-fact that `getCancelables` was exported for its registration part but not `close` part.

Reply:
That is say `cancelables.close` is supposed to be private to `StreamTask` but exported due to type but not on purpose, IIUC.

Reply:
> Not sure if I fully understand the concern. Can you help me understand more? Previously it is possible that the logic in stopMailboxProcessor is run in a different thread from the mailbox thread, e.g. the SourceThread in SourceStreamTask. So supposedly the stopMailboxProcessor logic should already be thread safe. And expose it to subclasses should also be safe. Is that the correct?

yes, it's correct.

But previously that was implementation detail of the `cancel()` method. By promoting this code to a separate `protected` method, you are making it a part of the `StreamTask` interface. That's the part I wasn't sure about. But I believe if you fix the first issue with not calling `cancelables.close()`, this change here can be safely reverted.